File: py/BST_print_level_order_in_lines.py
from collections import deque

class Node(object):
  def __init__(self, val):
    self.val = val
    self.left = None
    self.right = None

  def __str__(self):
    q = deque([None, self])
    str = ''

    while len(q):
      node = q.pop()

      if node and node.left:
        q.appendleft(node.left)

      if node and node.right:
        q.appendleft(node.right)

      if node:
        str += node.val

      if node is None and len(q):
        q.appendleft(None)
        str += '\n'

    return str


tree = Node('a')
tree.left = Node('b')
tree.right = Node('c')
tree.left.left = Node('d')
tree.left.right = Node('e')
tree.right.left = Node('f')
tree.right.right = Node('g')

print(tree)

File: BST_print_level_order_in_lines.js

class Node {
    constructor(val) {
        this.val = val
        this.left = null
        this.right = null
    }
}

function print_level_order_in_lines (root) {
    let levelOrder = ''
    const queue = [null, root]

    while (queue.length) {
        const node = queue.pop()

        if (node?.left) {
            queue.unshift(node.left)
        }

        if (node?.right) {
            queue.unshift(node.right)
        }

        levelOrder += node?.val || ''
        
        if (node === null && queue.length) {
          levelOrder += '\n'
          queue.unshift(node)
        }
    }

    return levelOrder
}


rootNode = new Node('a')

rootNode.left = new Node('b')
rootNode.right = new Node('c')

rootNode.left.left = new Node('d')
rootNode.left.right = new Node('e')
rootNode.right.left = new Node('f')
rootNode.right.right = new Node('g')

rootNode.left.left.left = new Node('h')
rootNode.left.left.right = new Node('i')
rootNode.left.right.left = new Node('j')
rootNode.left.right.right = new Node('k')
rootNode.right.left.left = new Node('l')
rootNode.right.left.right = new Node('m')
rootNode.right.right.left = new Node('n')
rootNode.right.right.right = new Node('o')

rootNode.left.left.left.left = new Node('p')
rootNode.left.left.left.right = new Node('q')
rootNode.left.left.right.left = new Node('r')
rootNode.left.left.right.right = new Node('s')
rootNode.left.right.left.left = new Node('t')
rootNode.left.right.left.right = new Node('u')
rootNode.left.right.right.left = new Node('v')
rootNode.left.right.right.right = new Node('w')
rootNode.right.left.left.left = new Node('x')
rootNode.right.left.left.right = new Node('y')
rootNode.right.left.right.left = new Node('z')

rootNode.left.left.left.left.left = new Node('A')
rootNode.left.left.left.left.right = new Node('B')
rootNode.left.left.left.right.left = new Node('C')
rootNode.left.left.left.right.right = new Node('D')
rootNode.left.left.right.left.left = new Node('E')
rootNode.left.left.right.left.right = new Node('F')

const res = print_level_order_in_lines(rootNode)

console.log('output:', res)





